browser/util.go

```diff
@@ -0,0 +1,111 @@
+package browser
+
+import (
+	"os"
+	"path/filepath"
+	"sync"
+)
+
+// Locking paths to make sure there's no interference
+
+var pathLock sync.Mutex
+
+var isLocked = make(map[string]bool)
+
+func LockPath(path string) {
+	pathLock.Lock()
+	isLocked[path] = true
+	pathLock.Unlock()
+}
+
+func UnlockPath(path string) {
+	pathLock.Lock()
+	isLocked[path] = false
+	pathLock.Unlock()
+}
+
+func IsLocked(path string) bool {
+	pathLock.Lock()
+	defer pathLock.Unlock()
+	if isLocked[path] {
+		return true
+	}
+	dir, f := filepath.Split(path)
+	for f != "" {
+		if isLocked[dir] {
+			return true
+		}
+		dir, f = filepath.Split(dir)
+	}
+	return false
+}
+
+// On exit make browser wait until every copy operation
+// is complete
+var op sync.WaitGroup
+
+func OpAdd() {
+	op.Add(1)
+}
+
+func OpDone() {
+	op.Done()
+}
+
+// Utility functions
+func WaitForOperationsToComplete() {
+	op.Wait()
+}
+
+func IsDir(dir string) bool {
+	file, err := os.Open(dir)
+	defer file.Close()
+	if err != nil {
+		return false
+	}
+	finfo, err := file.Stat()
+	if err != nil {
+		return false
+	}
+	return finfo.IsDir()
+}
+
+func ValidateDirPath(dir string) (string, bool) {
+	cleanDir := filepath.Clean(dir)
+	if !filepath.IsAbs(cleanDir) {
+		return cleanDir, false
+	}
+	return cleanDir, true
+}
+
+// Helper functions to wrap methods
+
+func OnePathWrapper(fun func(string, chan<- *ResultSet, string)) func(Command, chan<- *ResultSet) {
```

Comment:
This is a nit, but just to make the code cleaner how about implementing this as:
```go
RegisterSinglePathCommand("ls", List)
```
instead of:
```go
RegisterCommand("ls", OnePathWrapper(List))
```
The helper function `RegisterSinglePathCommand` would replace `OnePathWrapper`. Effectively, `RegisterSinglePathCommand` would just create a wrapper and then call `RegisterCommand`.

Note: it's quite possible that we should call it `registerSinglePathCommand` instead, so that we don't export it.
That's up to you on whether to export or not, just remember to document exported functions.